### ISSUE TITLE
bug: failed to specialize `drop` for generic argument of type `u32`

### ISSUE THREAD

**Issue by 0xNeshi** [starkware-libs/cairo#6321]
# Bug Report

**Cairo version:**

scarb 2.8.0+nightly-2024-08-31 (19fe5e032 2024-08-31)
cairo: 2.8.0 (ac0e135fb)
sierra: 1.6.0

Note that the issue occurs on Scarb v2.6.4, v2.8.0 and v2.8.1. It does not occur on v2.7.*.

**Current behavior:**

Given a generic list-like structure using `Felt252Dict` like:

```cairo
#[derive(Destruct)]
struct List<T, +Drop<T>, +Felt252DictValue<T>> {
    items: Felt252Dict<T>,
    len: u32,
}
```

Then instantiating said structure with values of signed integer, unsigned integer or felt, results in this error:

```console
thread 'scarb compile drop_bug-t45lkkfl5igrc' panicked at /cargo/git/checkouts/cairo-f086c7e6d4098a68/ac0e135/crates/cairo-lang-sierra-generator/src/utils.rs:335:9:
Failed to specialize: `drop<drop_bug::List::<core::integer::u32, core::integer::u32Drop, core::integer::U32Felt252DictValue>>`. Error: Could not specialize libfunc `drop` with generic_args: [[6]]. Error: Provided generic argument is unsupported.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at scarb/src/ops/compile.rs:163:14:
Compiler thread has panicked.: Any { .. }
error: `scarb metadata` exited with error
``` 

**Expected behavior:**

The compiler should be able to handle numeric types for generic arguments in the above struct.

**Steps to reproduce:**

- use any Scarb 2.8.* version
- create a new Cairo project locally
- copy/paste code from "Related code" section below into your _src/lib.cairo_ file
- run `scarb cairo-run` in your terminal
- error appears

**Related code:**

```cairo
use core::dict::Felt252Dict;

#[derive(Destruct)]
struct List<T, +Drop<T>, +Felt252DictValue<T>> {
    items: Felt252Dict<T>,
    len: u32,
}

#[generate_trait]
pub impl ListImpl<T, +Drop<T>, +Felt252DictValue<T>> of ListTrait<T> {
    fn new() -> List<T> {
        List { items: Default::default(), len: 0 }
    }

    fn length(self: @List<T>) -> u32 {
        *self.len
    }
}

fn main() {
    let list = ListTrait::<u32>::new();
    assert!(list.length() == 0, "Not empty");
    println!("All good");
}
```

**Other information:**

I noticed that other people had similar problems in the past (see https://github.com/starkware-libs/cairo/issues/2462 and https://github.com/starkware-libs/cairo/issues/2135), but since this was very long time ago, it doesn't seem the issues are related to this. As mentioned before, the problem reported in this issue occurs with Cairo 2.6.4 and 2.8.*.

---

Also, if instead of `u32`, the List is instantiated with `ByteArray` like:

```cairo
let list = ListTrait::<ByteArray>::new();
// ...
```

Then the following error occurs:

```console
error[E0002]: Method `length` could not be called on type `drop_bug::List::<core::byte_array::ByteArray, core::byte_array::ByteArrayDrop, ImplVarId(65)>`.
Candidate `ListTrait::length` inference failed with: Trait has no implementation in context: drop_bug::ListTrait::<core::byte_array::ByteArray, core::byte_array::ByteArrayDrop, ?1>.
 --> /home/nenad/repos/drop_bug/src/lib.cairo:22:18
    assert!(list.length() == 0, "Not empty");
                 ^****^

error: could not compile `drop_bug` due to previous error
error: `scarb metadata` exited with error
```

I'm not certain whether I simply forgot to include some impl trait as a constraint

---

Another related problem: let's say we remove the generic argument and hardcode the `u32` type. A similar problem _still_ occurs, but _only_ on Scarb v2.8.*:

```cairo
use core::dict::Felt252Dict;

#[derive(Destruct)]
struct List {
    items: Felt252Dict<u32>,
    len: u32,
}

#[generate_trait]
pub impl ListImpl of ListTrait {
    fn new() -> List {
        List { items: Default::default(), len: 0 }
    }

    fn length(self: @List) -> u32 {
        *self.len
    }
}

fn main() {
    let list = ListTrait::new();
    assert!(list.length() == 0, "Not empty");
    println!("All good");
}
```

Running this results in:

```console
thread 'scarb compile drop_bug-t45lkkfl5igrc' panicked at /cargo/git/checkouts/cairo-f086c7e6d4098a68/ac0e135/crates/cairo-lang-sierra-generator/src/utils.rs:335:9:
Failed to specialize: `drop<drop_bug::List>`. Error: Could not specialize libfunc `drop` with generic_args: [[6]]. Error: Provided generic argument is unsupported.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at scarb/src/ops/compile.rs:163:14:
Compiler thread has panicked.: Any { .. }
error: `scarb metadata` exited with error
```

**Comment by orizi** [comment]
Currently - support for impl parameters for `struct`s and `enum`s is highly limited.

Made the `Drop` dependency as a part of the impls, not a part of the `struct`.

**Comment by orizi** [comment]
```
struct List<T, +Drop<T>, +Felt252DictValue<T>> {
    items: Felt252Dict<T>,
    len: u32,
}
impl ListDestruct<T, Drop<T>> of Destruct<List<T>> {
      fn destruct(self: List<T>) nopanic {
           let List { items: _, len: _ } = self;
      }
}
```

**Comment by orizi** [comment]
The second part seems like an optimisation bug - looking into it.

**Comment by orizi** [comment]
that is fixed by the attached PR - will release 2.8.2 soon to solve.

**Comment by 0xNeshi** [comment]
Oh now I understand what you meant with the [last comment](https://github.com/starkware-libs/cairo/issues/6321#issuecomment-2323224082). Ok, waiting on the release!
